test/framework/machine_helpers.go

```diff
@@ -21,32 +21,66 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"github.com/pkg/errors"
 
+	"github.com/pkg/errors"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+type GetMachinesByClusterInput struct {
+	Lister      Lister
+	ClusterName string
+	Namespace   string
+}
+
+// GetMachinesByCluster returns the Machine objects that matches provided options for a cluster.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMachinesByCluster(ctx context.Context, input GetMachinesByClusterInput, options ...client.ListOption) []clusterv1.Machine {
+	opts := append(byClusterOptions(input.ClusterName, input.Namespace), options...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, opts...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
+// deploymentNameMachineOptions returns a set of ListOptions that allows to get all machine objects belonging to a machine deployment.
+func deploymentNameMachineOptions(deploymentName string) []client.ListOption {
+	return []client.ListOption{
+		client.MatchingLabels{clusterv1.MachineDeploymentLabelName: deploymentName},
+	}
+}
+
+// GetMachinesByClusterAndDeployment returns the Machine objects for a cluster belonging to a machine deployment.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMachinesByClusterAndDeployment(ctx context.Context, deploymentName string, input GetMachinesByClusterInput) []clusterv1.Machine {
+	return GetMachinesByCluster(ctx, input, deploymentNameMachineOptions(deploymentName)...)
+}
+
 // WaitForMachinesToBeUpgradedInput is the input for WaitForMachinesToBeUpgraded.
 type WaitForMachinesToBeUpgradedInput struct {
-	Lister                   Lister
-	Cluster                  *clusterv1.Cluster
-	KubernetesUpgradeVersion string
-	MachineCount             int
+	Lister         Lister
+	Cluster        *clusterv1.Cluster
+	UpgradeVersion string
+	MachineCount   int
+	MachineOptions []client.ListOption
```

Comment:
MachineOptions seems a bit confusing to me and I wasn't able to connect it to a slice of ListOption when used, would it be a bit more intuitive for users if we just called it ListOptions?
```suggestion
	ListOptions []client.ListOption
```